File: src/main/java/biz/lermitage/oga/util/IgnoreListTools.java
package biz.lermitage.oga.util;

import biz.lermitage.oga.Dependency;
import biz.lermitage.oga.DependencyState;
import biz.lermitage.oga.cfg.DefinitionMigration;
import biz.lermitage.oga.cfg.IgnoreItem;
import biz.lermitage.oga.cfg.IgnoreList;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;

/**
 * Ignore List tools.
 *
 * @author Jonathan Lermitage
 */
public class IgnoreListTools {

    private IgnoreListTools() {
    }

    public static boolean shouldIgnoreGroupId(
        Optional<IgnoreList> ignoreList,
        Dependency oldDep,
        DefinitionMigration newDep
    ) {
        if (ignoreList.isPresent()) {
            List<IgnoreItem> items = ignoreList.get().getIgnoreList();
            if (newDep.getState() == DependencyState.MIGRATED) {
                if (items != null) {
                    for (IgnoreItem ignoreItem : items) {
                        if (ignoreItem.isGroupIdOnly() && (ignoreItem.getGroupId().equals(oldDep.getGroupId()) || ignoreItem.getGroupId().equals(newDep.getNewerGroupId()))) {
                            return true;
                        }
                    }
                }
                return false;
            } else {
                if (items != null) {
                    for (IgnoreItem ignoreItem : items) {
                        if ((ignoreItem.isGroupIdOnly() && ignoreItem.getGroupId().equals(oldDep.getGroupId())) || shouldIgnoreProposal(ignoreList, newDep)) {
                            return true;
                        }
                    }
                }
                return false;
            }
        }
        return false;
    }

    public static boolean shouldIgnoreArtifactId(
        Optional<IgnoreList> ignoreList,
        Dependency oldDep,
        DefinitionMigration newDep
    ) {
        if (ignoreList.isPresent()) {
            List<IgnoreItem> items = ignoreList.get().getIgnoreList();
            if (newDep.getState() == DependencyState.MIGRATED) {
                if (items != null) {
                    for (IgnoreItem ignoreItem : items) {
                        if (ignoreItem.isGroupIdOnly()) {
                            if (ignoreItem.getGroupId().equals(oldDep.getGroupId()) || ignoreItem.getGroupId().equals(newDep.getNewerGroupId())) {
                                return true;
                            }
                        } else {
                            if ((ignoreItem.getGroupId().equals(oldDep.getGroupId()) && ignoreItem.getArtifactId().equals(oldDep.getArtifactId()))
                                || (ignoreItem.getGroupId().equals(newDep.getNewerGroupId()) && ignoreItem.getArtifactId().equals(newDep.getNewerArtifactId()))) {
                                return true;
                            }
                        }
                    }
                }
                return false;
            } else {
                if (items != null) {
                    for (IgnoreItem ignoreItem : items) {
                        if (ignoreItem.isGroupIdOnly()) {
                            if (ignoreItem.getGroupId().equals(oldDep.getGroupId()) || newDep.getProposedGroupIds().contains(ignoreItem.getGroupId())) {
                                return true;
                            }
                        } else {
                            if ((ignoreItem.getGroupId().equals(oldDep.getGroupId()) && ignoreItem.getArtifactId().equals(oldDep.getArtifactId())) || shouldIgnoreProposal(ignoreList, newDep)) {
                                return true;
                            }
                        }
                    }
                }
                return false;
            }
        }
        return false;
    }

    private static boolean shouldIgnoreProposal(Optional<IgnoreList> ignoreList, DefinitionMigration newDep) {
        if (ignoreList.isPresent()) {
            List<IgnoreItem> items = ignoreList.get().getIgnoreList();
            if (items != null) {
                List<String> ignoredDeps = new ArrayList<>();
                for (IgnoreItem ignoreItem : items) {
                    ignoredDeps.add(ignoreItem.getItem());
                }
                return new HashSet<>(ignoredDeps).containsAll(newDep.getProposal());
            }
        }
        return false;
    }
}

File: src/main/java/biz/lermitage/oga/util/IgnoreListTools.kt
package biz.lermitage.oga.util

import biz.lermitage.oga.Dependency
import biz.lermitage.oga.DependencyState
import biz.lermitage.oga.cfg.DefinitionMigration
import biz.lermitage.oga.cfg.IgnoreList
import java.util.Optional

/**
 * Ignore List tools.
 *
 * @author Jonathan Lermitage
 */
object IgnoreListTools {

    fun shouldIgnoreGroupId(
        ignoreList: Optional<IgnoreList>,
        oldDep: Dependency,
        newDep: DefinitionMigration
    ): Boolean {
        if (ignoreList.isPresent) {
            if (newDep.state == DependencyState.MIGRATED) {
                return ignoreList.get().ignoreList?.any { ignoreItem ->
                    ignoreItem.isGroupIdOnly && (ignoreItem.groupId == oldDep.groupId || ignoreItem.groupId == newDep.newerGroupId)
                } == true
            } else {
                return ignoreList.get().ignoreList?.any { ignoreItem ->
                    (ignoreItem.isGroupIdOnly && ignoreItem.groupId == oldDep.groupId) || shouldIgnoreProposal(ignoreList, newDep)
                } == true
            }
        }
        return false
    }

    fun shouldIgnoreArtifactId(
        ignoreList: Optional<IgnoreList>,
        oldDep: Dependency,
        newDep: DefinitionMigration
    ): Boolean {
        if (ignoreList.isPresent) {
            if (newDep.state == DependencyState.MIGRATED) {
                return ignoreList.get().ignoreList?.any { ignoreItem ->
                    if (ignoreItem.isGroupIdOnly) {
                        ignoreItem.groupId == oldDep.groupId || ignoreItem.groupId == newDep.newerGroupId
                    } else {
                        (ignoreItem.groupId == oldDep.groupId && ignoreItem.artifactId == oldDep.artifactId)
                            || (ignoreItem.groupId == newDep.newerGroupId && ignoreItem.artifactId == newDep.newerArtifactId)
                    }
                } == true
            } else {
                return ignoreList.get().ignoreList?.any { ignoreItem ->
                    if (ignoreItem.isGroupIdOnly) {
                        ignoreItem.groupId == oldDep.groupId || newDep.proposedGroupIds.contains(ignoreItem.groupId)
                    } else {
                        ((ignoreItem.groupId == oldDep.groupId && ignoreItem.artifactId == oldDep.artifactId) || shouldIgnoreProposal(ignoreList, newDep))
                    }
                } == true
            }
        }
        return false
    }

    private fun shouldIgnoreProposal(ignoreList: Optional<IgnoreList>, newDep: DefinitionMigration): Boolean {
        if (ignoreList.isPresent) {
            val ignoredDeps = ignoreList.get().ignoreList?.map { ignoreItem -> ignoreItem.item }?.toList()
            if (ignoredDeps?.containsAll(newDep.proposal!!) == true) {
                return true
            }
        }
        return false
    }
}
